File: java/core/src/main/java/com/whylogs/core/metrics/RegressionMetrics.java
package com.whylogs.core.metrics;

import com.google.common.base.Preconditions;
import com.whylogs.core.message.RegressionMetricsMessage;
import java.util.Map;
import java.util.Objects;
import lombok.Getter;
import lombok.NonNull;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import lombok.val;

@RequiredArgsConstructor
@Getter
@Slf4j
public class RegressionMetrics {
  @NonNull private final String predictionField;
  @NonNull private final String targetField;
  private double sumAbsDiff;
  private double sumDiff;
  private double sum2Diff;
  private long count;

  public void track(Map<String, ?> columns) {
    val prediction = (Number) columns.get(predictionField);
    val target = (Number) columns.get(targetField);

    final double diff = prediction.doubleValue() - target.doubleValue();
    this.sumAbsDiff += Math.abs(diff);
    this.sumDiff += diff;
    this.sum2Diff += diff * diff;
    this.count++;
  }

  public RegressionMetrics copy() {
    val res = new RegressionMetrics(this.predictionField, this.targetField);
    res.sumAbsDiff = this.sumAbsDiff;
    res.sumDiff = this.sumDiff;
    res.sum2Diff = this.sum2Diff;
    res.count = this.count;
    return res;
  }

  public RegressionMetrics merge(RegressionMetrics other) {
    if (other == null) {
      return this.copy();
    }

    Preconditions.checkState(
        Objects.equals(this.predictionField, other.predictionField),
        "Mismatched prediction fields: %s vs %s",
        this.predictionField,
        other.predictionField);
    Preconditions.checkState(
        Objects.equals(this.targetField, other.targetField),
        "Mismatched target fields: %s vs %s",
        this.targetField,
        other.targetField);

    val result = new RegressionMetrics(this.predictionField, this.targetField);
    result.sumAbsDiff = this.sumAbsDiff + other.sumAbsDiff;
    result.sumDiff = this.sumDiff + other.sumDiff;
    result.sum2Diff = this.sum2Diff + other.sum2Diff;
    result.count = this.count + other.count;

    return result;
  }

  public RegressionMetricsMessage.Builder toProtobuf() {
    return RegressionMetricsMessage.newBuilder()
        .setPredictionField(this.predictionField)
        .setTargetField(this.targetField)
        .setSumAbsDiff(this.sumAbsDiff)
        .setSumDiff(this.sumDiff)
        .setSum2Diff(this.sum2Diff)
        .setCount(this.count);
  }

  public static RegressionMetrics fromProtobuf(RegressionMetricsMessage msg) {
    if (msg == null || msg.getSerializedSize() == 0) {
      return null;
    }

    if ("".equals(msg.getPredictionField()) || "".equals(msg.getTargetField())) {
      log.warn("Skipping Regression metrics: prediction or target field not set");
      return null;
    }

    val res = new RegressionMetrics(msg.getPredictionField(), msg.getTargetField());
    res.sumAbsDiff = msg.getSumAbsDiff();
    res.sumDiff = msg.getSumDiff();
    res.sum2Diff = msg.getSum2Diff();
    res.count = msg.getCount();
    return res;
  }
}

File: python/whylogs/core/model_performance_metrics/regression_metrics.py
import math
from typing import List

from whylogs.core.proto.v0 import RegressionMetricsMessage


class RegressionMetrics:
    def __init__(self):
        self.count = 0
        self.sum_abs_diff = 0.0
        self.sum_diff = 0.0
        self.sum2_diff = 0.0

    def add(self, predictions: List[float], targets: List[float]):
        """
        Function adds predictions and targets computation of regression metrics.

        Args:
            predictions (List[float]):
            targets (List[float]):
        """

        # need to vectorize this
        for idx, target in enumerate(targets):
            self.sum_abs_diff += abs(predictions[idx] - target)
            self.sum_diff += predictions[idx] - target
            self.sum2_diff += (predictions[idx] - target) ** 2
            # To add later
            # self.nt_diff.track(predictions[idx] - target)
            self.count += 1

    def mean_absolute_error(self):
        if self.count == 0:
            return None
        return self.sum_abs_diff / self.count

    def mean_squared_error(self):
        if self.count == 0:
            return None
        return self.sum2_diff / self.count

    def root_mean_squared_error(self):
        if self.count == 0:
            return None
        return math.sqrt(self.sum2_diff / self.count)

    def merge(self, other):
        """
        Merge two seperate regression metrics.

        Args:
              other : regression metrics to merge with self
        Returns:
              RegressionMetrics: merged regression metrics
        """
        if other is None:
            return self

        if self.count == 0:
            return other
        if other.count == 0:
            return self

        new_reg = RegressionMetrics()
        new_reg.count = self.count + other.count
        new_reg.sum_abs_diff = self.sum_abs_diff + other.sum_abs_diff
        new_reg.sum_diff = self.sum_diff + other.sum_diff
        new_reg.sum2_diff = self.sum2_diff + other.sum2_diff

        return new_reg

    def to_protobuf(
        self,
    ):
        """
        Convert to protobuf

        Returns:
            TYPE: Protobuf Message
        """

        return RegressionMetricsMessage(
            prediction_field="0",
            target_field="0",
            count=self.count,
            sum_abs_diff=self.sum_abs_diff,
            sum_diff=self.sum_diff,
            sum2_diff=self.sum2_diff,
        )

    @classmethod
    def from_protobuf(
        cls,
        message: RegressionMetricsMessage,
    ):
        if message.ByteSize() == 0:
            return None

        reg_met = RegressionMetrics()
        reg_met.count = message.count
        reg_met.sum_abs_diff = message.sum_abs_diff
        reg_met.sum_diff = message.sum_diff
        reg_met.sum2_diff = message.sum2_diff

        return reg_met
